poly/schnorr.go

```diff
@@ -0,0 +1,350 @@
+	}
+	if s.partials[ps.Index] != nil {
+		return errors.New(fmt.Sprintf("A Partial Signature has already been added for this index %d", ps.Index))
+	}
+	if err := s.verifyPartialSig(ps); err != nil {
+		return errors.New(fmt.Sprintf("Partial signature to add is not valid : %v", err))
+	}
+	s.partials[ps.Index] = ps
+	return nil
+}
+
+// SchnorrSig  will generate the global schnorr signature
+// By reconstructing the secret that the partial responses contains
+func (s *Schnorr) SchnorrSig() (*SchnorrSig, error) {
+	// automatic verification
+	// TODO : change this into a bool flag or public method ?
+	if err := s.verify(); err != nil {
+		return nil, err
+	}
+
+	pri := PriShares{}
+	pri.Empty(s.info.Suite, s.info.T, s.info.N)
+	for i, ps := range s.partials {
+		pri.SetShare(ps.Index, *s.partials[i].Part)
+	}
+	// lagrange interpolation to compute the gamma
+	gamma := pri.Secret()
+	sig := &SchnorrSig{
+		random:    s.random.Pub,
+		signature: &gamma,
+	}
+	return sig, nil
+}
+
+// VerifySchnorrSig will verify if a given signature is correct regarding the message
+// NOTE: This belongs to the schnorr structs however it can be called at any time you want
+// This check is static, meaning it only needs the longterm shared secret, and the signature to
+// check. Think of the schnorr signature as a black box having two inputs
+//  - a message to be signed + a random secret ==> NewRound
+//  - a message + a signature to check on ==> VerifySchnorrSig
+func (s *Schnorr) VerifySchnorrSig(sig *SchnorrSig, msg []byte) error {
+	// gamma * G
+	left := s.info.Suite.Point().Mul(s.info.Suite.Point().Base(), *sig.signature)
+
+	randomCommit := sig.random.SecretCommit()
+	publicCommit := s.Longterm.Pub.SecretCommit()
+	hash, err := s.hashMessage(msg, randomCommit)
+
+	if err != nil {
+		return err
+	}
+	// RandomSecretCOmmit + H( ...) * LongtermSecretCommit
+	right := s.info.Suite.Point().Add(randomCommit, s.info.Suite.Point().Mul(publicCommit, hash))
+
+	if !left.Equal(right) {
+		return errors.New("Signature could not have been verified against the message")
+	}
+	return nil
+}
+
+//// DECODING / ENCODING /////
+// Decoding :
+// Use the Schnorr struct to produce already initialized struct
+// 	then call Unmarshal on them. Theses functions are in Schnorr
+// so you can not forget to init the struct with the right fields.
+// Encoding is simple as calling MarshalBinary on them
+
+// Wrapper function to help for the decoding
+func (s *Schnorr) EmptyPartialSig() *PartialSchnorrSig {
+	return new(PartialSchnorrSig).Init(s.info.Suite)
+}
+func (s *Schnorr) EmptySchnorrSig() *SchnorrSig {
+	return new(SchnorrSig).Init(s.info)
+}
+
+// Init is needed before decoding a PartialSchnorrSig so it can now about the length of a secret
+func (pss *PartialSchnorrSig) Init(suite abstract.Suite) *PartialSchnorrSig {
+	pss.Suite = suite
+	return pss
+}
+
+// Marshal a PartialSchnorrSig into a packet
+func (pss *PartialSchnorrSig) MarshalBinary() ([]byte, error) {
```

Comment:
Are you sure you even need these custom MarshalBinary/UnmarshalBinary methods? At least for PartialSchnorrSig, it looks to me like the generic abstract.Read and abstract.Write functions should work fine (after eliminating the abstract.Suite element from PartialSchnorrSig, which your comment says was only needed for marshal/unmarshal anyway).